### PR TITLE
Lpfg 173  line manager notified when required learning done

### PR DESCRIPTION
**Pull request by smoga-git**

### CODE REVIEW

File: src/main/java/uk/gov/cslearning/record/service/xapi/action/CompletedAction.java [hunk 1]
```diff
@@ -25,6 +25,7 @@
 
     @Override
     public void replay(CourseRecord courseRecord, ModuleRecord moduleRecord) {
+        System.out.println("Reached");
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/xapi/action/CompletedAction.java]: remove/change to log

File: src/main/java/uk/gov/cslearning/record/service/xapi/activity/Activity.java [hunk 2]
```diff
@@ -27,7 +27,7 @@
         Set<Class<? extends Activity>> classes = reflections.getSubTypesOf(Activity.class);
         for (Class<? extends Activity> clazz : classes) {
             try {
-                clazz.getDeclaredConstructor(Statement.class).newInstance(null);
+           //     clazz.getDeclaredConstructor(Statement.class).newInstance(null);
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/xapi/activity/Activity.java]: think this can be uncommented

File: src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java [hunk 3]
```diff
@@ -43,6 +45,21 @@ public LearnerRecordController(ActivityRecordService activityRecordService, User
...
+    @Autowired
+    private LearningJob learningJob;
+
+    @GetMapping(path = "/notify")
+    public ResponseEntity notifyme() {
+        try {
+            learningJob.sendNotificationForCompletedLearning();
+        } catch (Exception e) {
+            System.out.println(e);
+
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java]: Handle error appropriately

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java [hunk 4]
```diff
@@ -54,6 +60,8 @@
 
     private UserRecordService userRecordService;
 
+    private static final String COMPLETED = "COMPLETED";
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: Move statics to top of class

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java [hunk 5]
```diff
@@ -64,7 +72,89 @@ public LearningJob(UserRecordService userRecordService, IdentityService identity
         this.notificationRepository = notificationRepository;
     }
 
-    @Transactional
+    private void CheckAndNotifyLineManager(CivilServant civilServant, Identity identity,Course course, LocalDateTime completedDate) throws NotificationClientException {
+
+        if (civilServant.getLineManagerEmail() == null) {
+            LOGGER.error("User has no line manager!");
+        } else {
+            Boolean sendMail = false;
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: I think you can use the lower-case, primitive `boolean` here. It will have lower memory overheads

File: src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java (hunk 3 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java]: Opps , changed to LOGGER.error

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java (hunk 4 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: done

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java (hunk 5 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: okay sure

File: src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java [hunk 6]
```diff
@@ -43,6 +45,21 @@ public LearnerRecordController(ActivityRecordService activityRecordService, User
         return new ResponseEntity<>(new Records(records), OK);
     }
 
+    @Autowired
+    private LearningJob learningJob;
+
+    @GetMapping(path = "/notify")
```
[the remark this answers is not in the document]
Comment (m24murray) [reply src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java]: We're doing something similar with the required learning notifications - we can expose it with profiles. Let's chat up later

File: environ/ci [hunk 7]
```diff
@@ -18,3 +19,22 @@ docker tag cshr/${SERVICE_NAME}:latest cshr/${SERVICE_NAME}:${BRANCH_TAG}-${TRAV
...
+docker tag cshr/${SERVICE_NAME}:latest cshr/${SERVICE_NAME}:${BRANCH_TAG}
+docker tag cshr/${SERVICE_NAME}:latest cshr/${SERVICE_NAME}:${BRANCH_TAG}-${TRAVIS_BUILD_NUMBER}
+docker tag cshr/${SERVICE_NAME}:latest cshr/${SERVICE_NAME}:${COMMIT_TAG}
+
+<<<<<<< HEAD
+docker push cshr/lpg-learner-record
+>>>>>>> 7cf473a... Updating CI.
+=======
+docker push cshr/${SERVICE_NAME}
+>>>>>>> fab5d43... Missing configuration
```
Comment (m24murray) [inline environ/ci]: Merge conflict text here

File: src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java [hunk 8]
```diff
@@ -22,6 +27,8 @@
 
     private UserRecordService userRecordService;
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(LearnerRecordSummaryController.class);
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java]: Wrong class name in your logger

File: environ/ci (hunk 7 above)
Comment (smoga-git) [reply environ/ci]: fixed

File: src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java (hunk 8 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cslearning/record/api/LearnerRecordController.java]: fixed

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java [hunk 9]
```diff
@@ -64,6 +72,82 @@ public LearningJob(UserRecordService userRecordService, IdentityService identity
...
+        }
+
+    }
+    public void sendNotificationForCompletedLearning() throws NotificationClientException {
+
+        Collection<Identity> identities = identityService.listAll();
+
+        for (Identity identity : identities) {
+
+            LOGGER.info("Got identity with uid {} ({})", identity.getUsername(), identities.size());
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: `Got identity with uid identity-manager@domain.com (6)`

maybe reword this or don't show size?

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java [hunk 10]
```diff
@@ -64,6 +72,82 @@ public LearningJob(UserRecordService userRecordService, IdentityService identity
         this.notificationRepository = notificationRepository;
     }
 
+     void CheckAndNotifyLineManager(CivilServant civilServant, Identity identity,Course course, LocalDateTime completedDate) throws NotificationClientException {
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: lower case 'C'

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java (hunk 9 above)
Comment (smoga-git) [reply src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: Fair enough. Made it clearer that that is a total.

File: src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java [hunk 11]
```diff
@@ -64,6 +72,82 @@ public LearningJob(UserRecordService userRecordService, IdentityService identity
...
+                                    } catch (NotificationClientException nce) {
+                                        LOGGER.error("Error notifying line manger about course completion");
+                                        throw nce;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    throw new HttpClientErrorException(HttpStatus.UNPROCESSABLE_ENTITY);
```
Comment (m24murray) [inline src/main/java/uk/gov/cslearning/record/service/scheduler/LearningJob.java]: Just because an civil servant doesnt exist in CSRS, doesnt mean we should want to throw an exception here. Might it be worth just logging this and passing over?
